internal/resources/webform/webform.js

```
@@ -2354,6 +2404,115 @@ window.initGRPCForm = function(services, invokeURI, metadataURI, debug) {
+        history = history.slice(0, maxHistory - 1);
+        history.unshift(item);
+        writeHistory();
+        updateHistoryUI();
+    }
+
+    const updateHistoryUI = () => {
+        const list = $('#grpc-history-list');
+        list.empty();
+        const accordion = $('<div>');
+        list.append(accordion);
+        for (let i = 0; i < history.length; i++) {
+            const item = history[i];
+            const id = `grpc-history-item-${i}`;
+            const dataString = JSON.stringify(item.request.data, null, 4);
+            const valid = services[item.service] && services[item.service].includes(item.method);
+            accordion.append(`<div class="historyItemHeader" id="${id}">
+                <span><button class="delete" id="delete-${id}">X</button></span>
+                <span class="historyItemTime">${new Date(item.time).toLocaleString()}</span>
+                <span class="historyItemMethod">${item.service}.${item.method}</span>
+                <span class="historyItemResult">${item.result}</span>
+                <span class="historyItemLoad">
+                    <button class="load" ${valid ? '' : 'disabled'} id="load-${id}">Load</button>
+                    ${valid ? '' : '<span class="invalidHistory">Service or method no longer available</span>'}
+                </span>
+            </div>`);
+            accordion.append(`<div class="historyItemPanel">
+                <div>
+                    <span>Request:</span>
+                    <span><pre id="json">${dataString.slice(0, 250)}${dataString.length > 250 ? '...' : ''}</pre></span>
+                </div>
+                <div>
+                    <span>MetaData:</span>
+                    <span>${JSON.stringify(item.request.metadata)}</span>
+                </div>
+            </div>`);
+            $(`#delete-${id}`).click((evt) => {
+                deleteHistoryItem(i);
+                evt.preventDefault();
+                evt.stopImmediatePropagation();
```

What is the purpose of these? Since they are just buttons, and not submit inputs, I didn't think anything would happen on click. Maybe these are unnecessary and can be removed?

---

It prevents the onClick of the accordion header from occurring when pressing the button specifically.

---

Ah, I see. So if we let it propagate, the accordion component would try to collapse? How about a comment to that effect?